feat(set): add basic set membership operations

gforsyth opened the pull request:
Adds `is_in` and `index_in` functions to determine whether a value is a
member of a given set of values, and the index of that value if it
exists, respectively.

I put these in a new yaml file to start with, but I think we could
_probably_ stick them in `comparisons`?  Open to thoughts on where these
belong.  If we are planning to add other `set` operations (union,
difference, intersection, etc) then we may be better served by having a
separate file.

gforsyth commented on extensions/functions_set.yaml:
```diff
@@ -0,0 +1,89 @@
+    impls:
+      - args:
+          - value: i8
+            name: "x"
```
this is not a great name, but `value` and `values` seem too easily confused

gforsyth commented on extensions/functions_set.yaml:
```diff
@@ -0,0 +1,89 @@
+            name: "x"
+          - value: i8
+            name: "values"
+        return: int64?
```
I have all of these returning a nullable int64 which I think makes sense

jacques-n commented:
I haven't looked at the details here but saw the commit message. `is_in` doesn't need a function, the orlist expression type already provides this functionality, I belive.

jacques-n commented on extensions/functions_set.yaml:
```diff
@@ -0,0 +1,89 @@
+            name: "values"
+        return: bool
+  -
+    name: "index_in"
```
I'd expect these to either be variadic or have an argument that is a list of items. I'd also tend to think these should be defined generically e.g. 

```
- args: 
  - value: T
  - value: List<T>
```

cpcloud replied:
This looks to be resolved.

gforsyth commented:
> haven't looked at the details here but saw the commit message. `is_in` doesn't need a function, the orlist expression type already provides this functionality, I belive.

Thanks, I thought that might be the case but erred on the side of inclusion.  Removed and updated.

cpcloud commented on extensions/functions_set.yaml:
```diff
@@ -0,0 +1,15 @@
+    description: >
+      Checks the membership of a value in a set of values
+
+      Returns the 0-based index value of some input `T` if it is present,
```
Thanks for clarifying the index base :)

I think there are a couple things to resolve in the description:

1. What does `is present` mean? I'll suggest that it means the needle element is equal to (according substrait's `equals` function) an element in the haystack list.
2. What is the behavior when the needle is `NULL` (regardless of whether there's a `NULL` in the `list`). I believe as a consequence of point 1 it'll be `NULL` since `NULL = NULL` returns `NULL`.

cpcloud replied:
There's also `NaN` too ... yay?

gforsyth replied:
`NaN` -> ... `NULL` ?

cpcloud replied:
Gosh, this is really a fraught issue. See https://github.com/cockroachdb/cockroach/issues/18860 for example.

It seems like the decision to make `nan = nan` be `true` is from postgres's decision:

> in order to allow numeric values to be sorted and used in tree-based indexes, PostgreSQL treats NaN values as equal, and greater than all non-NaN values.

Many other incumbent databases also behave this way.

Pros of IEEE semantics:

1. We don't violate expectations of people who aren't steeped in SQL practicalities.

Pros of not-IEEE semantics:

1. NaNs (can) have a well-defined ordering
1. NaNs can be used in indexes

Thoughts?

cpcloud reviewed (changes requested):
Looks mostly good, just needs a bit of clarification in the description around what it means for an element to be present.

jvanstraten reviewed (changes requested):
Nullability should be set to `DECLARED_OUTPUT`, or the `?` flag is ignored (at least by my interpretation) since the default is `MIRROR`.

Presumably this will return the index of the *first* element that matches. After all, `list` allows for duplicates. Or is it allowed to return any matching index (better for parallelization)? The description should be explicit about this.

LGTM otherwise.

gforsyth commented:
For now I've opted to declare that it returns the first matching value's index.  We can extend the functionality later (returning all of the matching indices?).
And I've added the nullability declaration
